cmd/catalog/main.go

```diff
@@ -106,60 +109,59 @@ func main() {
 		*catalogNamespace = catalogNamespaceEnvVarValue
 	}
 
-	var useTLS bool
-	if *tlsCertPath != "" && *tlsKeyPath == "" || *tlsCertPath == "" && *tlsKeyPath != "" {
-		logger.Warn("both --tls-key and --tls-crt must be provided for TLS to be enabled, falling back to non-https")
-	} else if *tlsCertPath == "" && *tlsKeyPath == "" {
-		logger.Info("TLS keys not set, using non-https for metrics")
-	} else {
-		logger.Info("TLS keys set, using https for metrics")
-		useTLS = true
-	}
-
-	// Serve a health check.
-	healthMux := http.NewServeMux()
-	healthMux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+	mux := http.NewServeMux()
+	profile.RegisterHandlers(mux)
+	mux.Handle("/metrics", promhttp.Handler())
+	mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusOK)
 	})
 
-	// Serve profiling if enabled
-	if *profiling {
-		logger.Infof("profiling enabled")
-		profile.RegisterHandlers(healthMux)
+	s := http.Server{
+		Handler: mux,
+		Addr:    ":8080",
 	}
+	listenAndServe := s.ListenAndServe
 
-	go http.ListenAndServe(":8080", healthMux)
-
-	metricsMux := http.NewServeMux()
-	metricsMux.Handle("/metrics", promhttp.Handler())
-	if useTLS {
-		tlsGetCertFn, err := filemonitor.OLMGetCertRotationFn(logger, *tlsCertPath, *tlsKeyPath)
+	if *tlsCertPath != "" && *tlsKeyPath != "" {
+		logger.Info("TLS keys set, using https for metrics")
+		getCertificate, err := filemonitor.OLMGetCertRotationFn(logger, *tlsCertPath, *tlsKeyPath)
 		if err != nil {
 			logger.Errorf("Certificate monitoring for metrics (https) failed: %v", err)
 		}
 
-		go func() {
-			httpsServer := &http.Server{
-				Addr:    ":8081",
-				Handler: metricsMux,
-				TLSConfig: &tls.Config{
-					GetCertificate: tlsGetCertFn,
-				},
-			}
-			err := httpsServer.ListenAndServeTLS("", "")
+		s.Addr = ":8443"
+		s.TLSConfig = &tls.Config{
+			GetCertificate: getCertificate,
+			// GetConfigForClient: func(_ *tls.ClientHelloInfo) (*tls.Config, error) {
+			// 	// todo: rotate
+			// 	return nil, nil
+			// },
+		}
+		if *clientCAPath != "" {
+			pem, err := ioutil.ReadFile(*clientCAPath)
 			if err != nil {
-				logger.Errorf("Metrics (https) serving failed: %v", err)
+				logger.Fatalf("failed to load ca bundle: %v", err)
 			}
-		}()
+			pool := x509.NewCertPool()
+			pool.AppendCertsFromPEM(pem)
+			s.TLSConfig.ClientCAs = pool
+			s.TLSConfig.ClientAuth = tls.VerifyClientCertIfGiven
+		}
+		listenAndServe = func() error {
+			return s.ListenAndServeTLS("", "")
+		}
+	} else if *tlsCertPath != "" || *tlsKeyPath != "" {
+		logger.Fatal("both --tls-key and --tls-crt must be provided for TLS to be enabled")
 	} else {
-		go func() {
-			err := http.ListenAndServe(":8081", metricsMux)
-			if err != nil {
-				logger.Errorf("Metrics (http) serving failed: %v", err)
-			}
-		}()
+		logger.Info("TLS keys not set, using non-https for metrics")
 	}
```

Comment:
This code reappears in [olm/main.go](https://github.com/operator-framework/operator-lifecycle-manager/pull/2225/files#diff-52ebf4e1c25c87743e2d504e8015be1767d6140aadf638304142cdbb54216b3eR139-R171), can we create a function shared between the two?